### PR TITLE
Rewrite wallet sessions with persistent state

### PR DESCRIPTION
**Pull request by stevenroose**
I think we better merge the error PR first so that in this one, I can return GDK error codes when a user logs in but the wallet doesn't exist yet.

(Or do we want to silently register when a login is done?)

### CODE REVIEW

**Comment by greenaddress** [comment]
register in this case should be a no op IMHO

File: src/wallet.rs [hunk 1]
```diff
@@ -69,89 +79,172 @@ pub struct Wallet {
...
+        let mnem = Mnemonic::from_phrase(&mnemonic[..], Language::English).map_err(Error::Bip39)?;
+        let (master_xpriv, external_xpriv, internal_xpriv) = Wallet::calc_seeds(&mnem);
+        let fp = hex::encode(master_xpriv.fingerprint(&SECP).as_bytes());
+
+        // create the wallet in Core
+        let tmp_rpc = network.connect(None)?;
+        match tmp_rpc.create_wallet(fp.as_str(), Some(true))?.warning {
+            None => {}
+            Some(ref s) if s.is_empty() => {}
+            Some(warning) => warn!(
```
Comment (shesek) [inline src/wallet.rs:196]: What kind of warnings can `createwallet` return? Should this not be considered an error that fails the operation?

File: src/wallet.rs [hunk 2]
```diff
@@ -52,15 +54,23 @@ impl AddressMeta {
     }
 }
```
Comment (shesek) [inline src/wallet.rs:76]: Maybe add a general comment describing the use of address labels for storage?

File: src/wallet.rs [hunk 3]
```diff
@@ -69,89 +79,172 @@ pub struct Wallet {
...
 
 impl Wallet {
-    pub fn new(network: &'static Network) -> Result<Self, Error> {
-        let rpc = network.connect(None)?;
-        Ok(Wallet {
-            network,
+    /// Get the address to use to store persistent state.
+    fn persistent_state_address(
```
Comment (shesek) [inline src/wallet.rs:103]: Perhaps `persistent_state_address()` and `load_persistent_state()` are better suited as methods of `PersistentWalletState`?

File: src/wallet.rs (hunk 1 above)
Comment (stevenroose) [reply src/wallet.rs:196]: `createwallet` returns an error if it's not successful. I've seen that happen :)

This is the only occurrence of the warning I could find:
```c++
        if (r == BerkeleyEnvironment::VerifyResult::RECOVER_OK)
        {
            warningStr = strprintf(_("Warning: Wallet file corrupt, data salvaged!"
                                     " Original %s saved as %s in %s; if"
                                     " your balance or transactions are incorrect you should"
                                     " restore from a backup."),
                                   walletFile, backup_filename, walletDir);
        }
        if (r == BerkeleyEnvironment::VerifyResult::RECOVER_FAIL)
        {
            errorStr = strprintf(_("%s corrupt, salvage failed"), walletFile);
            return false;
        }
```

Comment (stevenroose) [reply src/wallet.rs:196]: So it seems like it should be safe to ignore it, but if it happens often, something is up.

File: src/wallet.rs (hunk 2 above)
Comment (stevenroose) [reply src/wallet.rs:76]: I added a commit with this.

File: src/wallet.rs (hunk 3 above)
Comment (stevenroose) [reply src/wallet.rs:103]: Hmm, I was thinking of `PersistentWalletState` as just a structure of data. I would have defined it inside the method if it wasn't that we can't share a struct across methods without defining its own type.

**Pull request merged**